File: sdk/typescript/src/pipeline.ts
/** Composite pipeline APIs (Rust-backed). */

import {
  buildSkillNodeCatalogNative,
  classifyAndCountCatalogNative,
  coordinateBm25PruneNative,
  pruneCatalogBm25AndRetrieveNative,
  searchSkillsAndSelectNative,
} from "./native.js";
import { PolicyContextNative } from "./native.js";
import type {
  ClassifyAndCountCatalogResult,
  JsonRecord,
  SkillNodeCatalogItem,
} from "./types.js";

type PolicyContext = InstanceType<typeof PolicyContextNative>;

export interface PruneBm25Options {
  scoreTool?: number;
  scoreToolEnum?: number;
  pruneEnums?: boolean;
  pipeline?: string[];
}

export interface SearchSkillsOptions {
  threshold?: number;
  maxTokens?: number;
  frontmatterUpperLimit?: number;
  itemKind?: string;
}

function pruneOptionsToNative(
  options?: PruneBm25Options,
): JsonRecord | undefined {
  if (!options) {
    return undefined;
  }
  const out: JsonRecord = {};
  if (options.scoreTool !== undefined) {
    out.score_tool = options.scoreTool;
  }
  if (options.scoreToolEnum !== undefined) {
    out.score_tool_enum = options.scoreToolEnum;
  }
  if (options.pruneEnums !== undefined) {
    out.prune_enums = options.pruneEnums;
  }
  if (options.pipeline !== undefined) {
    out.pipeline = options.pipeline;
  }
  return Object.keys(out).length > 0 ? out : undefined;
}

function searchOptionsToNative(
  options?: SearchSkillsOptions,
): JsonRecord | undefined {
  if (!options) {
    return undefined;
  }
  const out: JsonRecord = {};
  if (options.threshold !== undefined) {
    out.threshold = options.threshold;
  }
  if (options.maxTokens !== undefined) {
    out.max_tokens = options.maxTokens;
  }
  if (options.frontmatterUpperLimit !== undefined) {
    out.frontmatter_upper_limit = options.frontmatterUpperLimit;
  }
  if (options.itemKind !== undefined) {
    out.item_kind = options.itemKind;
  }
  return Object.keys(out).length > 0 ? out : undefined;
}

export function pruneCatalogBm25AndRetrieve(
  catalogData: JsonRecord,
  buildCatalog: JsonRecord,
  catalogIndex: JsonRecord,
  query: string,
  scoringCtx: PolicyContext,
  outputCtx: PolicyContext,
  options?: PruneBm25Options,
): JsonRecord {
  return pruneCatalogBm25AndRetrieveNative(
    catalogData,
    buildCatalog,
    catalogIndex,
    query,
    scoringCtx,
    outputCtx,
    pruneOptionsToNative(options),
  ) as JsonRecord;
}

export function classifyAndCountCatalog(
  catalogData: JsonRecord,
  tools?: JsonRecord[] | null,
): ClassifyAndCountCatalogResult {
  return classifyAndCountCatalogNative(
    catalogData,
    tools ?? undefined,
  ) as ClassifyAndCountCatalogResult;
}

export function searchSkillsAndSelect(
  entries: JsonRecord[],
  query: string,
  options?: SearchSkillsOptions,
): JsonRecord {
  return searchSkillsAndSelectNative(
    entries,
    query,
    searchOptionsToNative(options),
  ) as JsonRecord;
}

export function buildSkillNodeCatalog(
  entries: JsonRecord[],
): SkillNodeCatalogItem[] {
  return buildSkillNodeCatalogNative(entries) as SkillNodeCatalogItem[];
}

export interface CoordinateBm25Options {
  skills?: SearchSkillsOptions;
  tools?: PruneBm25Options;
}

function coordinateOptionsToNative(
  options?: CoordinateBm25Options,
): JsonRecord | undefined {
  if (!options) {
    return undefined;
  }
  const out: JsonRecord = {};
  const skills = searchOptionsToNative(options.skills);
  const tools = pruneOptionsToNative(options.tools);
  if (skills) {
    out.skills = skills;
  }
  if (tools) {
    out.tools = tools;
  }
  return Object.keys(out).length > 0 ? out : undefined;
}

export function coordinateBm25Prune(
  skillsEntries: JsonRecord[],
  catalogData: JsonRecord,
  buildCatalog: JsonRecord,
  catalogIndex: JsonRecord,
  query: string,
  scoringCtx: PolicyContext,
  outputCtx: PolicyContext,
  options?: CoordinateBm25Options,
): JsonRecord {
  return coordinateBm25PruneNative(
    skillsEntries,
    catalogData,
    buildCatalog,
    catalogIndex,
    query,
    scoringCtx,
    outputCtx,
    coordinateOptionsToNative(options),
  ) as JsonRecord;
}

File: src/cyt/indexer/pipeline.py
"""Composite pipeline APIs — re-exports cyt-indexer-sdk."""

from __future__ import annotations

from typing import Any

from cyt_indexer.pipeline import (
    build_skill_node_catalog as _build_skill_node_catalog,
)
from cyt_indexer.pipeline import (
    classify_and_count_catalog as _classify_and_count_catalog,
)
from cyt_indexer.pipeline import (
    coordinate_bm25_prune as _coordinate_bm25_prune,
)
from cyt_indexer.pipeline import (
    prune_catalog_bm25_and_retrieve as _prune_catalog_bm25_and_retrieve,
)
from cyt_indexer.pipeline import (
    search_skills_and_select as _search_skills_and_select,
)

from cyt.indexer.build import CatalogIndex
from cyt.pruners.policies import PolicyContext

__all__ = [
    "build_skill_node_catalog",
    "classify_and_count_catalog",
    "coordinate_bm25_prune",
    "prune_catalog_bm25_and_retrieve",
    "search_skills_and_select",
]


def build_skill_node_catalog(entries: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Batch-load rerankable node bodies (includes cached ``token_count`` when present)."""
    return _build_skill_node_catalog(entries)


def classify_and_count_catalog(
    catalog_data: dict[str, Any],
    tools: list[dict[str, Any]] | None = None,
) -> dict[str, Any]:
    """Classify optional catalog chunks and optionally count tool JSON tokens."""
    return _classify_and_count_catalog(catalog_data, tools)


def search_skills_and_select(
    entries: list[dict[str, Any]],
    query: str,
    *,
    options: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """BM25 skill search with optional frontmatter gate and greedy budget selection."""
    return _search_skills_and_select(entries, query, options=options)


def prune_catalog_bm25_and_retrieve(
    catalog_data: dict[str, Any],
    build_catalog: dict[str, Any],
    catalog_index: CatalogIndex | dict[str, Any],
    query: str,
    scoring_ctx: PolicyContext,
    output_ctx: PolicyContext,
    *,
    options: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Partition, BM25 score, recompose, and retrieve tools in one native call."""
    return _prune_catalog_bm25_and_retrieve(
        catalog_data,
        build_catalog,
        catalog_index,
        query,
        scoring_ctx,
        output_ctx,
        options=options,
    )


def coordinate_bm25_prune(
    skills_entries: list[dict[str, Any]],
    catalog_data: dict[str, Any],
    build_catalog: dict[str, Any],
    catalog_index: CatalogIndex | dict[str, Any],
    query: str,
    scoring_ctx: PolicyContext,
    output_ctx: PolicyContext,
    *,
    options: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Run skills BM25 search and tool BM25 prune in parallel."""
    return _coordinate_bm25_prune(
        skills_entries,
        catalog_data,
        build_catalog,
        catalog_index,
        query,
        scoring_ctx,
        output_ctx,
        options=options,
    )
